app/models/spotlight/resource.rb

```diff
@@ -74,72 +35,141 @@ def needs_provider?
-    end
+      def exhibit_specific_solr_data
+        exhibit.solr_data if exhibit
+      end
 
-    def batch_size
-      Spotlight::Engine.config.solr_batch_size
-    end
+      def spotlight_resource_metadata_for_solr
+        {
+          Spotlight::Engine.config.resource_global_id_field => (to_global_id.to_s if persisted?),
+          Spotlight::SolrDocument.resource_type_field => self.class.to_s.tableize
+        }
+      end
 
-    def exhibit_specific_solr_data
-      exhibit.solr_data if exhibit
+      def document_model
+        exhibit.blacklight_config.document_model if exhibit
+      end
     end
 
-    def spotlight_resource_metadata_for_solr
-      {
-        Spotlight::Resource.resource_global_id_field => (to_global_id.to_s if persisted?),
-        Spotlight::SolrDocument.resource_type_field => self.class.to_s.tableize
-      }
-    end
+    concerning :Indexing do
+      ##
+      # Index the result of {#to_solr} into the index in batches of {#batch_size}
+      #
+      # @return [Integer] number of records indexed
+      def reindex
+        benchmark "Reindexing #{self} (batch size: #{batch_size})" do
+          count = 0
 
-    ##
-    # @return an enumerator of all the indexable documents for this resource
-    def documents_to_index
-      return to_enum(:documents_to_index) unless block_given?
+          run_callbacks :index do
+            documents_to_index.each_slice(batch_size) do |batch|
+              add_to_index(batch)
+              count += batch.length
+            end
 
-      data = to_solr
-      return if data.blank?
+            count
+          end
+        end
+      end
 
-      data &&= [data] if data.is_a? Hash
+      protected
 
-      data.reject(&:blank?).each do |doc|
-        yield doc.reverse_merge(existing_solr_doc_hash(doc[unique_key]) || {})
+      def reindex_with_lock
+        with_lock do
+          yield
+        end
       end
-    end
 
-    ##
-    # Get any exhibit-specific metadata stored in e.g. sidecars, tags, etc
-    # This needs the generated solr document
-    def existing_solr_doc_hash(id)
-      document_model.new(unique_key => id).to_solr if document_model && id.present?
-    end
+      def reindex_with_logging
+        time_start = Time.zone.now
 
-    def unique_key
-      if document_model
-        document_model.unique_key.to_sym
-      else
-        :id
+        count = yield
+
+        time_end = Time.zone.now
+
+        update(indexed_at: Time.current,
+               last_indexed_estimate: documents_to_index.size,
+               last_indexed_count: count,
+               last_index_elapsed_time: time_end - time_start)
+      end
+
+      private
+
+      def blacklight_solr
+        @solr ||= RSolr.connect(connection_config)
+      end
+
+      def connection_config
+        Blacklight.connection_config
+      end
+
+      def batch_size
+        Spotlight::Engine.config.solr_batch_size
+      end
+
+      def add_to_index(batch)
```

Comment:
I can live with "add_to_index" but I think "save_to_index" or "write_to_index" would be clearer.